File: backend/app/utils/context_utils.py
import json
from typing import Optional, List
from pathlib import Path


def wrap_article_as_message(article_path: str, role: str = "user") -> Optional[dict]:
    """
    将给定的文章路径包装成message格式
    
    Args:
        article_path: 文章文件路径
        role: message角色，默认为"user"
    
    Returns:
        包含文章内容的message字典，如果文件不存在或读取失败则返回None
    """
    path = Path(article_path)
    if not path.exists():
        return None
    
    try:
        with open(path, "r", encoding="utf-8") as f:
            content = f.read()
        
        return {
            "role": role,
            "content": content
        }
    except Exception:
        return None


def parse_mentions(mentions_str: Optional[str]) -> List[dict]:
    """
    解析mentions JSON字符串为字典列表
    
    Args:
        mentions_str: JSON格式的mentions字符串
    
    Returns:
        解析后的字典列表，如果为None或解析失败则返回空列表
    """
    if not mentions_str:
        return []
    
    try:
        mentions = json.loads(mentions_str)
        return mentions if isinstance(mentions, list) else []
    except (json.JSONDecodeError, TypeError):
        return []


def append_attachments_to_user_text(user_text: str, absolute_paths: List[str]) -> str:
    """
    将已持久化附件的绝对路径追加到用户可见文本末尾，供模型与会话存档使用。

    格式示例::

        [Attached files — server cache]
        - D:/.../workspace/local_data/cache/doc.pdf
    """
    if not absolute_paths:
        return user_text
    lines = "\n".join(f"- {p}" for p in absolute_paths)
    block = f"[Attached files — server cache]\n{lines}"
    if not user_text.strip():
        return block
    return f"{user_text.rstrip()}\n\n{block}"


def build_user_message_with_mentions(text: str, mentions: List[dict]) -> str:
    """
    构建包含@mention名称的用户消息
    
    Args:
        text: 用户输入的文本
        mentions: mention字典列表
    
    Returns:
        格式化后的消息，格式为 "@article1, @article2\n\nuser text"
        如果没有mentions则直接返回text
    """
    if not mentions:
        return text
    
    mention_names = [f"@{mention.get('name', '')}" for mention in mentions if mention.get('name')]
    if not mention_names:
        return text
    
    mentions_line = ", ".join(mention_names)
    return f"{mentions_line}\n\n{text}"


def get_article_context_messages(mentions: List[dict]) -> List[dict]:
    """
    从mentions中提取文章内容并转换为消息列表
    
    Args:
        mentions: mention字典列表，每个字典包含parsed_path和name字段
    
    Returns:
        消息字典列表，每条消息包含文章名称作为标题和文章内容
    """
    context_messages = []
    
    for mention in mentions:
        parsed_path = mention.get("parsed_path")
        name = mention.get("name", "未命名文章")
        
        if not parsed_path:
            continue
        
        message = wrap_article_as_message(parsed_path)
        if message:
            # 添加文章名称作为标题
            header = f"# 参考文章: {name}\n\n"
            message["content"] = header + message["content"]
            context_messages.append(message)
    
    return context_messages

File: backend-rs/src/service/context_utils.rs
use serde_json::Value;

/// 从 mentions 列表中解析 article 引用，构建上下文消息
///
/// mentions: 引用字符串列表，格式如 `["article:xxx", "url:yyy"]`
/// 返回要插入到 LLM messages 中的上下文消息列表
pub fn get_article_context_messages(mentions: &[Value]) -> Vec<Value> {
    let mut messages = Vec::new();

    for mention in mentions {
        let mention_str = match mention.as_str() {
            Some(s) => s,
            None => continue,
        };
        if let Some(content) = resolve_mention(mention_str) {
            messages.push(serde_json::json!({
                "role": "user",
                "content": format!("[引用内容]:\n{}", content)
            }));
        }
    }

    messages
}

/// 解析单个 mention 引用，返回引用的文本内容
fn resolve_mention(mention: &str) -> Option<String> {
    if let Some(article_id) = mention.strip_prefix("article:") {
        return resolve_article(article_id);
    }
    if let Some(url) = mention.strip_prefix("url:") {
        return resolve_url_content(url);
    }
    None
}

/// 从 records 服务中读取指定 article_id 的内容
fn resolve_article(_article_id: &str) -> Option<String> {
    // TODO: 后续从 records 服务加载文章内容
    None
}

/// 从缓存目录读取已 fetch 的 URL 内容
fn resolve_url_content(_url: &str) -> Option<String> {
    // TODO: 后续从缓存或直接 fetch 获取内容
    None
}

/// 将已持久化附件的绝对路径追加到用户可见文本末尾，供模型与会话存档使用。
///
/// 格式示例：
///
/// ```text
/// [Attached files — server cache]
/// - D:/.../workspace/local_data/cache/doc.pdf
/// ```
pub fn append_attachments_to_user_text(user_text: &str, absolute_paths: &[String]) -> String {
    if absolute_paths.is_empty() {
        return user_text.to_string();
    }
    let lines: Vec<String> = absolute_paths.iter().map(|p| format!("- {p}")).collect();
    let block = format!("[Attached files — server cache]\n{}", lines.join("\n"));
    if user_text.trim().is_empty() {
        block
    } else {
        format!("{}\n\n{}", user_text.trim_end(), block)
    }
}

// ── Tests ────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // ── append_attachments_to_user_text ──────────────────────────────────

    #[test]
    fn test_append_attachments_empty_paths_returns_original() {
        let result = append_attachments_to_user_text("hello", &[]);
        assert_eq!(result, "hello");
    }

    #[test]
    fn test_append_attachments_single_path() {
        let paths = vec!["D:/cache/doc.pdf".to_string()];
        let result = append_attachments_to_user_text("hello", &paths);
        assert!(result.starts_with("hello"));
        assert!(result.contains("Attached files"));
        assert!(result.contains("doc.pdf"));
    }

    #[test]
    fn test_append_attachments_multiple_paths() {
        let paths = vec!["/tmp/a.pdf".to_string(), "/tmp/b.pdf".to_string()];
        let result = append_attachments_to_user_text("hello", &paths);
        assert!(result.contains("- /tmp/a.pdf"));
        assert!(result.contains("- /tmp/b.pdf"));
    }

    #[test]
    fn test_append_attachments_blank_user_text() {
        let paths = vec!["/tmp/doc.pdf".to_string()];
        let result = append_attachments_to_user_text("  \n  ", &paths);
        assert!(result.starts_with("[Attached files"));
        assert!(result.contains("doc.pdf"));
    }

    #[test]
    fn test_get_article_context_messages_empty() {
        let mentions: Vec<Value> = vec![];
        let result = get_article_context_messages(&mentions);
        assert!(result.is_empty());
    }

    #[test]
    fn test_get_article_context_messages_article() {
        // article:xxx 格式 — 目前 resolve_article 返回 None，所以结果应为空
        let mentions = vec![Value::String("article:doc-123".to_string())];
        let result = get_article_context_messages(&mentions);
        assert!(result.is_empty());
    }

    #[test]
    fn test_get_article_context_messages_url() {
        // url:xxx 格式 — 目前 resolve_url_content 返回 None
        let mentions = vec![Value::String("url:https://example.com".to_string())];
        let result = get_article_context_messages(&mentions);
        assert!(result.is_empty());
    }

    #[test]
    fn test_get_article_context_messages_unknown_prefix() {
        // 未知前缀的 mention 应被忽略
        let mentions = vec![Value::String("unknown:something".to_string())];
        let result = get_article_context_messages(&mentions);
        assert!(result.is_empty());
    }

    #[test]
    fn test_get_article_context_messages_non_string() {
        // 非字符串的 mention 应被忽略
        let mentions = vec![serde_json::json!(42)];
        let result = get_article_context_messages(&mentions);
        assert!(result.is_empty());
    }

    #[test]
    fn test_get_article_context_messages_mixed() {
        let mentions = vec![
            serde_json::json!("article:doc-1"),
            serde_json::json!("url:https://example.com"),
            serde_json::json!("invalid"),
            serde_json::json!(99),
        ];
        // 所有 resolve 目前都返回 None
        let result = get_article_context_messages(&mentions);
        assert!(result.is_empty());
    }
}
